### DISCUSSION TITLE
Are the contents of an `ItemList` and `ListItem` dependant on the property they are used in?

### DISCUSSION THREAD

**Discussion by Toromyx** [schemaorg/schemaorg#3411] [Q&A] [answered]
This question came up in relation to <https://schema.org/recipeInstructions>.
This property is defined to be one of these types:

- [CreativeWork](https://schema.org/CreativeWork)
- [ItemList](https://schema.org/ItemList)
- [Text](https://schema.org/Text)

It is also described to be "an ordered list with HowToStep and/or HowToSection items." This fact is not defined in the structured data.

So can an `ItemList` be described as what would be a "generic type" in some programming languages? So for example `ItemList<CreativeWork>` defines the list items to be of the type `CreativeWork`.

**Comment by rrlevering** [comment, marked as answer]
By current schema.org semantics, the typing of the objects in the ItemList can be anything like an old school Java List.  But in practice, they are going to be the things in the range of the property they are used in like you say.  As a person who works on schema.org validation most of the time, this is super annoying.

I personally think schema.org could benefit from a move to a generic type more like you suggest, but it would require some more horizontal changes to collection structures and perhaps some thought to how they relate to RDF containers.

**Comment by Toromyx** [comment]
Thanks for the answer.

For future reference, I also found <https://developers.google.com/search/docs/appearance/structured-data/recipe#recipe-properties> which defines
- `recipeInstructions` to be of type `HowToStep`, `HowToSection`, or `Text`
- `itemListElement` in `HowToSection` to be of type `HowToStep`
- `itemListElement` in `HowToStep` to be of type `HowToDirection` or `HowToTip`

**Reply by rrlevering** [reply]
Yes, but those are Google specific requirements for a particular feature.  Google defines more strict requirements that hopefully should be a subset of the full range of the schema.org requirements.
